diff --git a/3rdparty/ext_json_c/CMakeLists.txt b/3rdparty/ext_json_c/CMakeLists.txt
index 20dcab4a45..a5b98f4df6 100755
--- a/3rdparty/ext_json_c/CMakeLists.txt
+++ b/3rdparty/ext_json_c/CMakeLists.txt
@@ -1,8 +1,8 @@
 SET(PREFIX_ext_json_c "${EXTPREFIX}" )
 ExternalProject_Add( ext_json_c
     DOWNLOAD_DIR ${EXTERNALS_DOWNLOAD_DIR}
-    URL https://github.com/json-c/json-c/archive/json-c-0.15-20200726.tar.gz
-    URL_HASH SHA256=4ba9a090a42cf1e12b84c64e4464bb6fb893666841d5843cc5bef90774028882 
+    URL https://github.com/json-c/json-c/archive/refs/tags/json-c-0.16-20220414.tar.gz
+    URL_HASH SHA256=3ecaeedffd99a60b1262819f9e60d7d983844073abc74e495cb822b251904185 
 
     INSTALL_DIR ${PREFIX_ext_json_c}
     CMAKE_ARGS -DCMAKE_INSTALL_PREFIX=${PREFIX_ext_json_c} -DCMAKE_BUILD_TYPE=${GLOBAL_BUILD_TYPE} ${GLOBAL_PROFILE}
